TailLight/wmi.cpp

```diff
@@ -1,6 +1,93 @@
 #include "driver.h"
 
 
+
+VOID FakeBISTTimerProc(_In_ WDFTIMER timer) {
+    WDFDEVICE Device = static_cast<WDFDEVICE>(WdfTimerGetParentObject(timer));    
+    COLOR_CONTROL* pColorController = WdfObjectGet_COLOR_CONTROL(timer);
+    NTSTATUS status = SetFeatureColor(Device, pColorController->Colors[pColorController->RemainingColors]);
+    if (!NT_SUCCESS(status)) {
+        DPF_SOMETHING_FAILED(instance, something); }
+
+    if (pColorController->RemainingColors) {
+        pColorController->RemainingColors = pColorController->RemainingColors - 1;
+        
+        BOOLEAN timerInQueue = WdfTimerStart(timer, WDF_REL_TIMEOUT_IN_SEC(1));
+        NT_ASSERTMSG("Previous active timer overwritten", !timerInQueue);
+    }
+    else {
+        pColorController->RemainingColors = REMAINING_COLORS_COUNT;
+    }
+}
+
+NTSTATUS FakeBIST(_In_ WDFDEVICE Device) {
+/*++
+    
+ Routine Description: Performs a fake Built-In Self Test (BIST)
+
+--*/
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+
+    WDFTIMER* pTimer = &WdfObjectGet_DEVICE_CONTEXT(Device)->FakeBISTTimer;
+    auto pColorController = WdfObjectGet_COLOR_CONTROL(*pTimer);
+
+    // FakeBIST is called in an arbitrary thread. If more than one thread is in
+    // FakeBIST at the same time, then the call to SetFeatureColor will result
+    // in an invalid taillight color sequence. Simply return that the device is
+    // busy using an invalid COLOR_CONTROL::RemainingColors value to signify
+    // that the hardware is free. Otherwise we can queue up the request or 
+    // block. Both of those mechanisms would require notifying that the
+    // hardware is busy and cancelation.
+    // 
+    // TODO: Maybe a function to get the last BIST test result and time.
+    if (pColorController->RemainingColors != REMAINING_COLORS_COUNT) {
+        return STATUS_DEVICE_BUSY;
+    }
+
+    InterlockedDecrement((LONG*) & pColorController->RemainingColors);
```

Comment:
Are you trying to be thread-safe here?

If YES:
* Then the current logic is not safe. You'll need to instead use the `InterlockedDecrement` return-value in the if-check above to prevent a race condition.
* Alternatively, consider enable `AutomaticSerialization` for the timer, so that you avoid the need for making this function thread-safe in the first place.

If NO:
* Then please switch to regular variable decrement.

Reply:
In the interest of clarity: Doesn't InterlockedDecrement return the new value of the variable that was decremented? In short, shouldn't both the return value and the in/out variable Addend will be the same? InterlockedDecrement function generates an lfence followed by a locked dec so there must be a gap in my understanding. Please advise.

Reply:
If several threads are entering `FakeBIST` simultaneously, then the `pColorController->RemainingColors` variable might suddenly change due to the "other" thread calling `InterlockedDecrement`. This makes it unsafe to check the `pColorController->RemainingColors` value directly in if-checks.

If thread safety is needed then you should instead use the return-value of `InterlockedDecrement` in if-checks. This is guaranteed to simultaneously executing threads unique values for the number of remaining colors.

Reply:
Ah, I see. Should have looked out for that. In any case, timer config's AutomaticSerialization by default is set to true. Since the timer runs at passive level, the timer will be synchronized with FakeBISTTimerProc, which also uses shared references to the device context https://www.osr.com/nt-insider/2014-issue3/understanding-sync-scope-wdf-drivers/#:~:text=Sync%20Scope%20provides%20a%20simple%20serialization%20scheme%20that,your%20WDFQUEUE%2C%20WDFFILEOBJECT%2C%20and%20WDFREQUEST%20to%20be%20serialized.

Reply:
Thanks for explaining. Makes sense to me.